### ISSUE TITLE
Caret holding shift too long, or holding shift then typing a character after a space crashes Geany sometimes (race condition)

### ISSUE THREAD

**Issue by Poikilos** [geany/geany#3927]
I think I encountered this before, but didn't know how to reproduce it. I'm pretty sure it only happened before when I pressed shift. :edit: I confirmed it can happen while holding shift long enough even without typing a character:

## Environment
- OS: Linux Mint 21.3
- Geany 1.38 "Sulamar" (built on or after 2021-10-14) Using GTK+ v3.24.33 and GLib v2.72.4 runtime libraries

## Steps to reproduce issue
- Extract the BAS file from the attached zip: [DIM_TYPES.BAS.zip](https://github.com/user-attachments/files/16321882/DIM_TYPES.BAS.zip)
- open the BAS file in Geany
- Set highlighter to PASCAL
- click after print e (or the statement before it--I'm not sure how many work)
- press enter
- press ' (straight single quote)
- press spacebar
- press shift+6 (caret)

*If you don't press spacebar it doesn't happen

*If you wait too long to press caret, it doesn't happen :confused: 
- but if you wait too long to press caret (or type some other character) then press backspace then quickly enter a caret again, the crash does happen.

* If you wait too long to press caret, but are holding Shift, it *does* happen

## Related issues
- #2841 (not sure if related)
- Likely other "random crash" (or Segmentation fault or core dumped) issues without a known cause are the same as this, since the steps to reproduce are so arbitrary.

Console output:
```
(geany:504867): Tagmanager-WARNING **: 14:02:32.564: ignoring null tag in /home/owner/git/qjak/JAKSOUND.BAS(line: 4)
Segmentation fault (core dumped)
```
- "ignoring null tag" may not be relevant to this issue since #3144 doesn't mention a crash
  - and the crash doesn't happen until the steps above are followed.

**Comment by elextr** [comment]
- Can't reproduce OP with Geany 2.1 currentish git.  It is probably an issue with the pascal lexer processing basic code that it doesn't understand (but shouldn't crash).  Probably fixed by lexilla upgrade.

- Closed issue #2841 refers to #2813 which is "closed fixed" and was in Geany 2.0 but not relevant since its VTE not Scintilla.

- The tagmanager "random crash" has nothing to do with the VTE crash in #2813 nor likely the OP.  Its likely in ctags, did you use pascal on that basic file?

- #3144 notes its ctags upstream issue and reported there, but thats js not basic so not relevant.

It good you are looking at other reported issues but in this case the similarities do not indicate any connection between the causes.

**Comment by techee** [comment]
Also, the Pascal ctags parser used to crash (in 1.38 which you are using) because of https://github.com/geany/geany/issues/3127 which is fixed now. Since in this case you are setting Pascal filetype for a Basic file, the Pascal ctags parser will be very confused by the input (still, it shouldn't crash).

**Comment by techee** [comment]
> Also, the Pascal ctags parser used to crash (in 1.38 which you are using) because of https://github.com/geany/geany/issues/3127 which is fixed now. Since in this case you are setting Pascal filetype for a Basic file, the Pascal ctags parser will be very confused by the input (still, it shouldn't crash).

@Poikilos You could probably test if it's this problem by removing the word `type` from line 16 of the file you submitted and re-trying the steps above on the modified file. If it's this issue, you shouldn't be able to reproduce the crash when there's no `type` word in the file.

**Comment by elextr** [comment]
@Poikilos just out of interest why are you treating basic filetypes as pascal?

**Comment by Poikilos** [comment]
> @Poikilos just out of interest why are you treating basic filetypes as pascal?

Simply because there was not a BASIC nor VB highlighter, and IIRC many of the keywords are the same. But wait...I see now there *is* a **FreeBASIC** highlighter so I should use that.

@techee 
It sounds like you have it covered but I will try again with the wrong highlighter in the later version after testing with the code change you mentioned.

**Comment by Poikilos** [comment]
- uninstalled it
- installed the nightly repo...instructions were a bit strange here: <https://nightly.geany.org/debian/bookworm/> since they say to use buster. I tried changing the deb source to say "bookworm" since Ubuntu 22.04 Jammy is based on bookworm, but it couldn't find the right version of deps:
```
The following packages have unmet dependencies:
 geany-plugin-git-changebar : Depends: libgit2-1.5 (>= 1.5.0) but it is not installable
 geany-plugin-workbench : Depends: libgit2-1.5 (>= 1.5.0) but it is not installable
```
- so I changed it to "bullseye". I made a separate file called /etc/apt/sources.list.d/geany-nightly.list containing only `deb http://geany.nightlybuilds.org/debian/bullseye stable main`
- I didn't try buster like instructions at link above say, because that seemed wrong.
- apt update
- apt install geany geany-plugins
- still says 1.38-1, but problem is gone.
- I thought maybe the version was lower than actual because of a git tag, so I tried `apt policy geany` and `apt-cache policy geany` and both say the source is `http://archive.ubuntu.com/ubuntu jammy/universe amd64 Packages`

I can say the issue is solved but I'm confused about what I did wrong above in my apt configuration.
